### DISCUSSION TITLE
Problema upgrade Mosquitto 1.5.7 Broker 3.1.1 a ver 2x

### DISCUSSION THREAD

**Discussion by saviothecnic** [indomus/forum#880] [GENERALE - Discussioni generiche]
Salve ragazzi ho il mio raspberry Mosquitto 1.5.7 Broker 3.1.1

Con questi comandi 

wget http://repo.mosquitto.org/debian/mosquitto-repo.gpg.key
sudo apt-key add mosquitto-repo.gpg.key
sudo wget http://repo.mosquitto.org/debian/mosquitto-buster.list
sudo apt update
sudo apt upgrade -y

Mi ha aggiornato alla versione 2.0.12

Solo che non mi parte piu

Il files di config è questo io gli ho aggiunto listener 1883 o provato anche con listener 1883 0.0.0.0

Ma niente non mi va cosaltro devo vedere per farlo andare post upgrade ?

 Place your local configuration in /etc/mosquitto/conf.d/

 A full description of the configuration file is at
 /usr/share/doc/mosquitto/examples/mosquitto.conf.example

pid_file /var/run/mosquitto.pid

persistence true
persistence_location /var/lib/mosquitto/

log_dest topic
log_dest syslog
log_dest stdout
log_type error
log_type warning
log_type notice
log_type information

connection_messages true
log_timestamp true
allow_anonymous false
password_file /etc/mosquitto/pwfile
bind_address 192.168.1.20
port 1883
listener 1883
autosave_on_changes true
persistence_file mosquitto.db
include_dir /etc/mosquitto/conf.d

**Comment by snakuzzo** [comment]
Innanzi tutto correggi il post scrivendo correttamente e non con font e formattazione a caso, altrimenti dobbiamo eliminare il topic.
Ti posto di seguito il mio file di configurazione funzionante... magari adegua i path e controlla sul sito ufficiale la documentazione per escludere parametri che sono cambiati e/o deprecati.

```
persistence true
persistence_location /mosquitto/data
listener 1883
log_dest file /mosquitto/log/mosquitto.log
log_timestamp_format %Y-%m-%dT%H:%M:%S
allow_anonymous false
password_file /mosquitto/config/pwfile
```

**Reply by saviothecnic** [reply]
Scusami non so perche è uscito quel font fose dal copia e incolla dei codici d' erroe ha preso qualche formattazione strana
Ok provo a confrontar econ il tuo files questo è errore che ho io
![ErroreMosquitto](https://user-images.githubusercontent.com/6697807/137552705-041a26ba-84de-45b0-bbc9-251a6150ad71.jpg)

**Reply by snakuzzo** [reply]
così si capisce poco... prova a lanciarlo in foreground a mano...

`mosquitto -c /etc/mosquitto/mosquitto.conf`

(ovviamente se `/etc/mosquitto/mosquitto.conf`  è il tuo file di configurazione)

**Reply by saviothecnic** [reply]
Ok mi da questi tre errori 

pi@raspberrypi:~ $ mosquitto -c /etc/mosquitto/mosquitto.conf
1634333290: The 'bind_address' option is now deprecated and will be removed in a future version. The behaviour will default to true.
1634333290: The 'port' option is now deprecated and will be removed in a future version. Please use 'listener' instead.
1634333290: Error: Unable to write pid file.

**Reply by saviothecnic** [reply]
Non riesco a capire ho provato sia ad aggiungere listener
che listener 1883 0.0.0.0 ma dice che gira solo in modo locale dove sbaglio ?
![ErroreMosquitto2](https://user-images.githubusercontent.com/6697807/137557577-ecccfd41-1f1f-44f6-a035-7c420b3ae7d5.jpg)

**Comment by saviothecnic** [comment]
Ora ho ripristinato un immagine dove avevo la 1.5.7 cosi intanto lascio lavorare il sistema

Ma se mi date dritte domani riprovo ad upgradare alla versione 2.x e cercare di farlo andare

Se serve come info aggiuntiva questa e la risposta agli stessi comandi con la versione 1.5.7 che mi funziona

Posto anche la config della 1.5.7

Praticamente quello che cambia è la prima riga
pid_file /var/run/mosquitto.pid
che ho provato anche a commentare con cancelletto ma non so se è cosi che devo fare

![Mosquitto157](https://user-images.githubusercontent.com/6697807/137558765-6072039d-f31d-4305-be97-aeb84ff73114.jpg)

![ConfigMosquitto157](https://user-images.githubusercontent.com/6697807/137558998-19a00ecb-c05e-4b12-9d46-02d518cfee61.jpg)

e le due righe sotto che ho commentato con il cancelletto 
e provato a mettere al loro posto listener 1883 oppure listener 1883 0.0.0.0  
bind_address 192.168.1.20
port 1883

Ma senza esiti positivi :(

**Reply by snakuzzo** [reply]
Se guardi bene ti dice `address already in use` ... vuol dire che già hai un'istanza attiva che ha occupato quell'IP:PORTA.
Finché non fermi il servizio già attivi, non ti partirà mai.
Inoltre ti ho postato la mia configurazione perfettamente funzionante...parti da quella

**Comment by saviothecnic** [comment]
Scusa la mia ignoranza sul mondo linux
"alla fine avevo rimesso la vecchia immagine per andare avanti per questo non mi sono fatto sentire per un po"
Ma vorrei cercare d' aggiornare dalla 1.5.7 che ho ora all' ultima versione 2.x
Come faccio a capire e fermare i servizi attivi ?
Come vedi ho provato a metere la tua config commendando la mia con canceletto
![immagine](https://user-images.githubusercontent.com/6697807/147390013-9abbf92f-1d57-4215-8783-727d29f428f6.png)

Ma ho gli stessi errori quando dighito da terminale mosquitto

![immagine](https://user-images.githubusercontent.com/6697807/147390065-9f176e9e-5782-4be4-8d5c-1ec5f943386a.png)
Eppure ho messo esattamente la tua config che a te va senza problemi non capisco cosa non va nel mio sistema

Dimenticavo Buon Natale

**Reply by snakuzzo** [reply]
Ciao...Buon Natale anche a te ma ti ho già detto l'altra volta che se ti dice `Address already in use` vuol dire che hai ancora un servizio attivo sulla porta 1883 (e non c'entra niente il fatto che sia Linux...avresti avuto lo stesso messaggio su qualsiasi altro sistema operativo)
Hai probabilmente un altro servizio attivo (mosquitto?) che parte di default al boot.

Lancia questo comando e vedrai qual è il processo che è attivo su quella porta

`sudo lsof -i:1883`

**Comment by saviothecnic** [comment]
Ok ho risolto la config attuale è questa presa grazie agli spunti di questo forum 

![immagine](https://user-images.githubusercontent.com/6697807/147393109-9f1a299e-30ad-44e6-a203-a23a6977b8de.png)

Praticamente io lanciavo mosquitto per vedere se andava
Mentre dovevo lanciare sudo mosquitto -c /etc/mosquitto/mosquitto.conf

**Reply by snakuzzo** [reply]
ok. A posto

**Comment by saviothecnic** [comment]
Ansi no ho cantato vittoria troppo presto :( ho notato che riavviandolo avvolte parte e avvolte no
Verifico il mio files di avvio ora è cosi

![immagine](https://user-images.githubusercontent.com/6697807/147403248-8bbfba2b-19d4-43d1-86bf-92c7a640e5be.png)

Credo devo modificarlo cosi esatto ?
Domanda posso aggiungere qualcosa che se cade il servizio lo riattiva in automatico ?

![immagine](https://user-images.githubusercontent.com/6697807/147403288-fa0f415b-7f66-4b30-9bf7-b363f64b4aed.png)

**Reply by snakuzzo** [reply]
Ma ti pare normale fare gli screenshot degli script? Bah
Se proprio devi condividere per cercare di farti dare una mano prendi il codice e lo posti evidenziando cosa modifichi...non è che uno può mettersi lì a guardare le immagini e giocare a "trova l'intruso".
Detto ciò non capisco cosa stai facendo...le cose non si fanno in questo modo. Solitamente ogni servizio ha io suo file "service" ed è lì che si imposta quando farlo partire e dopo cosa.

**Reply by saviothecnic** [reply]
Scusatemi ma la prima volta ho postato il contenuto
Mi avete detto che aveva caretteri e from che davano problemi al forum 
Per questo sono passato alle immagini

Quindi non è il modo corretto aggiungere `sudo mosquitto -c /etc/mosquitto/mosquitto.conf` in rc.local ?
E dove e come dovrei farlo partire ?

**Reply by snakuzzo** [reply]
Quando installi mosquitto con `apt` ti viene creato un file `service`...e lo puoi anche vedere al termine dell'installazione (di seguito la parte finale)...

```
...
Unpacking libwebsockets16:armhf (4.0.20-2) ...
Selecting previously unselected package mosquitto.
Preparing to unpack .../5-mosquitto_2.0.11-1_armhf.deb ...
Unpacking mosquitto (2.0.11-1) ...
Setting up libmosquitto1:armhf (2.0.11-1) ...
Setting up libev4:armhf (1:4.33-1) ...
Setting up libcjson1:armhf (1.7.14-1) ...
Setting up libdlt2:armhf (2.18.6-1) ...
Setting up libwebsockets16:armhf (4.0.20-2) ...
Setting up mosquitto (2.0.11-1) ...
Created symlink /etc/systemd/system/multi-user.target.wants/mosquitto.service -> /lib/systemd/system/mosquitto.service.
Processing triggers for man-db (2.9.4-2) ...
Processing triggers for libc-bin (2.31-13+rpt2+rpi1) ...
```

Come puoi vedere troverai il file `/etc/systemd/system/multi-user.target.wants/mosquitto.service` ed è quello che verrà usato per avviare e/o fermare il servizio.

Utilizzando `systemctl` (o anche `service`) potrai gestire il servizio e verificare in che stato è. Sappi che di default, una volta installato, mosquitto verrà avviato ed abilitato all'autostart al boot del server.

Poi...per definire le dipendenze al boot puoi andare a modificare direttamente il file `.service`. Nello specifico, in questo caso, puoi giocare con i parametri `After` e `Wants. Se guardi il file ecco cosa puoi notare...

```
pi@raspberrypi:~ $ cat /etc/systemd/system/multi-user.target.wants/mosquitto.service
[Unit]
Description=Mosquitto MQTT Broker
Documentation=man:mosquitto.conf(5) man:mosquitto(8)
After=network.target
Wants=network.target

[Service]
Type=notify
...
```

Detto ciò, concludo invitandoti a leggere qui...
https://man7.org/linux/man-pages/man1/systemctl.1.html
https://man7.org/linux/man-pages/man5/systemd.service.5.html

**Reply by saviothecnic** [reply]
Grazie della delucidazione

Allora ho provato vedendo i tuoi comandi ed effettivamente se lancio systemctl
tra i vari servizi noto
`mosquitto.service     loaded failed     failed          Mosquitto MQTT Broker `
Quindi dice che il servizio del Broker è fallito pero non capisco 
perche da windows con MQTT explorer vedo che il server MQTT è su e la varie comunicazioni passano

Se invece lancio cat /etc/systemd/system/multi-user.target.wants/mosquitto.service
Ho questa risposta
`i@raspberrypi:~ $ cat /etc/systemd/system/multi-user.target.wants/mosquitto.service
[Unit]
Description=Mosquitto MQTT Broker
Documentation=man:mosquitto.conf(5) man:mosquitto(8)
After=network.target
Wants=network.target

[Service]
Type=notify
NotifyAccess=main
ExecStart=/usr/sbin/mosquitto -c /etc/mosquitto/mosquitto.conf
ExecReload=/bin/kill -HUP $MAINPID
Restart=on-failure
ExecStartPre=/bin/mkdir -m 740 -p /var/log/mosquitto
ExecStartPre=/bin/chown mosquitto /var/log/mosquitto
ExecStartPre=/bin/mkdir -m 740 -p /run/mosquitto
ExecStartPre=/bin/chown mosquitto /run/mosquitto

[Install]
WantedBy=multi-user.target
`
Quindi effettivamente lo lancia da qui non non devo averlo anche su rc.local esatto ?
Altrimenti lo avvio due volte ho capito bene ?
Pero se non lo metto in rc.local ho notato che spesso un riavvio su tre non mi parte come servizio

**Reply by snakuzzo** [reply]
Tu una volta installato non devi modificare nulla a mano. Come già ti ho detto ti crea il service e abilita in automatico l'autostart al boot.
Quindi tutto quello che hai fatto tu "a manella" dopo non serve/serviva a niente.
Io non posso sapere che casini hai fatto...lo sai solo tu. Quindi ti tocca armarti di pazienza e sistemare

**Comment by saviothecnic** [comment]
Allora ragionando forse ho capito dove sbaglio
Quando installai all' epoca Mosquitto 1.5.7

Seguii una guida che citava che alla fine per farlo partire in automatico all avviso
con `sudo nano /etc/rc.local` si doveva aggiungere la righa `sudo mosquitto –d prima di "exit 0`

Quindi quando ho upgradato e rimasta questa righa in rd.local
Credevo che si risolvesse sostituendola con s`udo mosquitto -c /etc/mosquitto/mosquitto.conf`
Ma in realta solo ora capsico che dovevo proprio levare questa righa essendo gia prensente in system

Forse sta cosa vi conviene aggiungerla anche nella guida d' upgrade che avete fatto
Perche per profani come me non è una cosa scontata da sapere che in quella prima si doveva aggiungere 
in rc.local e nella nuova è in automatico in system e quindi va rimossa la precendenre che si era messa a manina su rc.local

**Reply by snakuzzo** [reply]
Guarda...sinceramente non ricordo questa cosa che dici e comunque, nonostante la guida sia stata seguita da tantissimi, sei l'unico che ha riscontrato questa cosa.
Controlleremo, ma la guida credo vada bene così com'è.

**Comment by marcgrc** [comment]
Una domandina 
Posso usare un rasperry già attivo con mqtt brocker e visualizzare su Home Assistant i sonoff ?

**Reply by snakuzzo** [reply]
certo...se è sulla stessa LAN/VLAN ed è raggiungibile dai Sonoff e da HA puoi (anzi...devi) farlo

**Reply by marcgrc** [reply]
E come ?

**Reply by snakuzzo** [reply]
Scusa, ma perché questa domanda ?
Hai un broker MQTT a tua insaputa ?
